### PR TITLE
[ios] Change the bookmarks/tracks color from the color icon modally

### PR DESCRIPTION
**Pull request by kirylkaveryn**
This PR implements:
1. Modal presentation for the color selection
When a user want to quickly change a single option there is better solution to make the screen modal. Why modal? Apple's hig says:
> Help people perform a distinct, narrowly scoped task without **losing track of their previous context**

https://developer.apple.com/design/human-interface-guidelines/modality

2. Modal presentation for the category selection (to match with the color selection)
3. Possibility to change color by tapping on the `circle color icon` without opening the `Edit bookmark` screen (it's really faster to change color now!)
4. New reusable ColorPicker class that:
 - use native color picking for tracks if > iOS14.0
 - use old color picking for all other cases and `iPad designed for Mac` (because on mac this color picker have bad design)
 
Tested on: 
iPhone 11Pro ios17.2
iPhone 6 ios12.5
iPhone 15 pro 17.2 (sim)
iPad 5air 5 (sim)
iPad designed for Mac

Results:

Simulator Screen Recording - iPhone 15 Pro - 2024-03-01 at 14 04 54
Simulator Screen Recording - iPhone 15 Pro - 2024-03-01 at 14 05 06
Simulator Screen Recording - iPhone 15 Pro - 2024-03-01 at 14 33 37
Simulator Screen Recording - iPad Air (5th generation) - 2024-03-01 at 14 38 25

image
image

### CODE REVIEW

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm [hunk 1]
```diff
@@ -667,6 +667,21 @@ - (void)updateBookmark:(MWMMarkID)bookmarkId
     bookmark->SetCustomName(title.UTF8String);
 }
 
+- (void)updateBookmark:(MWMMarkID)bookmarkId
+              setColor:(MWMBookmarkColor)color {
```
Comment (biodranik) [inline iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:671]:
```suggestion
- (void)updateBookmark:(MWMMarkID)bookmarkId setColor:(MWMBookmarkColor)color {
```
nit: Prefer fewer lines of code if possible (if it fits 120-130 chars), it helps to review code faster.

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm [hunk 2]
```diff
@@ -667,6 +667,21 @@ - (void)updateBookmark:(MWMMarkID)bookmarkId
...
+  auto editSession = self.bm.GetEditSession();
+
+  auto bookmark = editSession.GetBookmarkForEdit(bookmarkId);
+  if (!bookmark)
```
Comment (biodranik) [inline iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:675]: Let's crash here.

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm [hunk 3]
```diff
@@ -667,6 +667,21 @@ - (void)updateBookmark:(MWMMarkID)bookmarkId
...
+    return;
+
+  auto kmlColor = convertBookmarkColor(color);
```
Comment (biodranik) [inline iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:678]: convertBookmarkColor is not clear. A better name hints what is converted to what.

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm [hunk 4]
```diff
@@ -697,6 +712,21 @@ - (void)updateTrack:(MWMTrackID)trackId
     track->SetName(title.UTF8String);
 }
 
+- (void)updateTrack:(MWMTrackID)trackId
+              setColor:(UIColor *)color {
```
Comment (biodranik) [inline iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:716]: ditto

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm [hunk 5]
```diff
@@ -697,6 +712,21 @@ - (void)updateTrack:(MWMTrackID)trackId
...
+    auto editSession = self.bm.GetEditSession();
+
+    auto track = editSession.GetTrackForEdit(trackId);
+    if (!track)
```
Comment (biodranik) [inline iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:720]: crash with explanation reason

File: iphone/Maps/UI/EditBookmark/EditBookmarkViewController.swift [hunk 6]
```diff
@@ -170,6 +167,21 @@ final class EditBookmarkViewController: MWMTableViewController {
     editingCompleted?(true)
     goBack()
   }
+
+  @objc private func openColorPicker() {
+    ColorPicker.shared.present(from: self, pickerType: .bookmarkColorPicker(bookmarkColor), completionHandler: { [weak self] color in
+      self?.bookmarkColor = BookmarkColor.bookmarkColor(from: color)
+      self?.tableView.reloadRows(at: [IndexPath(row: InfoSectionRows.color.rawValue, section: Sections.info.rawValue)],
+                           with: .none)
```
Comment (biodranik) [inline iphone/Maps/UI/EditBookmark/EditBookmarkViewController.swift:175]:
```suggestion
      self?.tableView.reloadRows(at: [IndexPath(row: InfoSectionRows.color.rawValue, section: Sections.info.rawValue)], with: .none)
```

**Review by biodranik — COMMENTED** [review]
Does it work well on iOS 12?

**Comment by kirylkaveryn** [comment]
> Does it work well on iOS 12?

Yes! 

https://github.com/organicmaps/organicmaps/assets/79797627/1438c8f0-02bb-4a22-b534-610379129feb

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm (hunk 2 above)
Comment (biodranik) [reply iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:674]: Will it work in release? Is there a better way to crash and show some error to a user?

```

Assertions are disabled if the preprocessor macro NS_BLOCK_ASSERTIONS is defined or the ENABLE_NS_ASSERTIONS Xcode build setting is disabled.
```

Comment (kirylkaveryn) [reply iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:674]: Should we show this internal error from the core to the user? And what message should we use?
There are no exception at all in the BookmarksManager.

Comment (biodranik) [reply iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:674]: Ok, I see that C++ core never checks for nullptr after getting bookmarks using the same method. So let's replace if here and in other similar places with
NSAssert(bookmark, @"Invalid bookmark id: @%llu", bookmarkId);

Comment (kirylkaveryn) [reply iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm]: done!

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm [hunk 7]
```diff
@@ -667,42 +660,75 @@ - (void)updateBookmark:(MWMMarkID)bookmarkId
...
+}
+
+- (void)moveBookmark:(MWMMarkID)bookmarkId toGroupId:(MWMMarkGroupID)groupId {
+  auto const currentGroupId = self.bm.GetBookmark(bookmarkId)->GetGroupId();
+  auto editSession = self.bm.GetEditSession();
+  if (groupId != kml::kInvalidMarkGroupId)
+    editSession.MoveBookmark(bookmarkId, currentGroupId, groupId);
 }
 
 - (void)updateTrack:(MWMTrackID)trackId
```
Comment (biodranik) [inline iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:687]: Does it make sense to introduce separate methods to move track, to rename it, and to change its color?

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm [hunk 8]
```diff
@@ -667,42 +660,75 @@ - (void)updateBookmark:(MWMMarkID)bookmarkId
...
+    editSession.MoveTrack(trackId, currentGroupId, groupId);
 
-    auto track = editSession.GetTrackForEdit(trackId);
-    if (!track)
-      return;
+  auto track = editSession.GetTrackForEdit(trackId);
+  if (!track) {
+    NSAssert(false, @"Invalid track id: @%llu", trackId);
+    return;
+  }
```
Comment (biodranik) [inline iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:700]:
```suggestion
  ASSERT(track, ("Invalid track id:", trackId));
```

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm [hunk 9]
```diff
@@ -667,42 +660,75 @@ - (void)updateBookmark:(MWMMarkID)bookmarkId
...
-        toGroupId:(MWMMarkGroupID)groupId {
-    auto const currentGroupId = self.bm.GetTrack(trackId)->GetGroupId();
-    auto editSession = self.bm.GetEditSession();
-    if (groupId != kml::kInvalidMarkGroupId)
-      editSession.MoveTrack(trackId, currentGroupId, groupId);
+- (void)updateTrack:(MWMTrackID)trackId setColor:(UIColor *)color {
+  auto editSession = self.bm.GetEditSession();
+
+  auto track = editSession.GetTrackForEdit(trackId);
+  if (!track) {
```
Comment (biodranik) [inline iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:715]: ditto

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm [hunk 10]
```diff
@@ -667,42 +660,75 @@ - (void)updateBookmark:(MWMMarkID)bookmarkId
...
+  auto const newColor = [MWMBookmarksManager getColorFromUIColor:color];
+
+  if (newColor != currentColor)
+    track->SetColor(newColor);
+}
+
+- (void)moveTrack:(MWMTrackID)trackId toGroupId:(MWMMarkGroupID)groupId {
+  auto const currentGroupId = self.bm.GetTrack(trackId)->GetGroupId();
+  auto editSession = self.bm.GetEditSession();
+  if (groupId != kml::kInvalidMarkGroupId)
```
Comment (biodranik) [inline iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:730]: ditto

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm (hunk 7 above)
Comment (kirylkaveryn) [reply iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:687]: For now, we have 3 edit cases:
- only change the color (from color icon)
- only change the group (from leading swipe action)
- change the color, the group and name from Edit bookmark screen by tapping `Save button`

So these three methods reflect all these actions.

It is not convenient to call
```
- (void)updateTrack:(MWMTrackID)trackId
         setGroupId:(MWMMarkGroupID)groupId
              color:(UIColor *)color
              title:(NSString *)title { 
```
when you want to change only one param.

And not convenient to call 3 methods when you want to update all at once by tapping Save.

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm [hunk 11]
```diff
@@ -667,42 +660,75 @@ - (void)updateBookmark:(MWMMarkID)bookmarkId
...
               color:(UIColor *)color
               title:(NSString *)title {
-    auto const currentGroupId = self.bm.GetTrack(trackId)->GetGroupId();
-    auto editSession = self.bm.GetEditSession();
-    if (groupId != kml::kInvalidMarkGroupId)
-      editSession.MoveTrack(trackId, currentGroupId, groupId);
+  auto const currentGroupId = self.bm.GetTrack(trackId)->GetGroupId();
+  auto editSession = self.bm.GetEditSession();
+  if (groupId != kml::kInvalidMarkGroupId)
+    editSession.MoveTrack(trackId, currentGroupId, groupId);
```
[the remark this answers is not in the document]
Comment (kirylkaveryn) [reply iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:694]: I added an additional check to skip unnessessary moving:
```
  if (currentGroupId != groupId)
    editSession.MoveBookmark(bookmarkId, currentGroupId, groupId);
```

**Comment by vng** [comment]
The commits set is strange, why is [planet] data here?

**Comment by kirylkaveryn** [comment]
> The commits set is strange, why is [planet] data here?

Fixed (I made a mistake while rebasing for the beta)

**Comment by kirylkaveryn** [comment]
ios 17.2 (simulator)
Simulator Screen Recording - iPad Air (5th generation) - 2024-03-07 at 20 59 59
Simulator Screen Recording - iPhone 15 Pro - 2024-03-07 at 21 11 51

**Pull request merged**